app/integration_action.go

```diff
@@ -250,6 +257,87 @@ func (a *App) DoActionRequest(rawURL string, body []byte) (*http.Response, *mode
 	return resp, nil
 }
 
+type LocalResponseWriter struct {
+	data   []byte
+	status int
+}
+
+func (w *LocalResponseWriter) Header() http.Header {
+	return make(http.Header)
+}
+
+func (w *LocalResponseWriter) Write(bytes []byte) (int, error) {
+	w.data = make([]byte, len(bytes))
+	copy(w.data, bytes)
+	return len(w.data), nil
+}
+
+func (w *LocalResponseWriter) WriteHeader(statusCode int) {
+	w.status = statusCode
+}
+
+func (a *App) DoLocalRequest(rawURL string, body []byte) (*http.Response, *model.AppError) {
```

Comment:
Notwithstanding my comments below, might we instead just reuse `app.ServePluginRequest`, constructing an `http.Request` with the incoming parameters?

Reply:
I made the changes for all review items so far but this one.  I followed what I believe you requested, though in testing the remind & matterpoll plugins, matterpoll is only 20% functional in terms of buttons.  remind worked fine.  Is the following what you had in mind?  If so I will make the changes and dive into why matterpoll isn't working with it.

```go
func (a *App) DoLocalRequest(rawURL string, body []byte) (*http.Response, *model.AppError) {
	rawURL = strings.TrimPrefix(rawURL, "/")
	inURL, err := url.Parse(rawURL)
	if err != nil {
		return nil, model.NewAppError("DoActionRequest", "api.post.do_action.action_integration.app_error", nil, "err="+err.Error(), http.StatusBadRequest)
	}
	result := strings.Split(inURL.Path, "/")
	if len(result) < 2 {
		return nil, model.NewAppError("DoActionRequest", "api.post.do_action.action_integration.app_error", nil, "err=Unable to find pluginId", http.StatusBadRequest)
	}
	if result[0] != "plugins" {
		return nil, model.NewAppError("DoActionRequest", "api.post.do_action.action_integration.app_error", nil, "err=plugins not in path", http.StatusBadRequest)
	}
	pluginId := result[1]

	path := strings.TrimPrefix(inURL.Path, "plugins/"+pluginId)

	w := LocalResponseWriter{}
	r, err := http.NewRequest("POST", path, bytes.NewReader(body))
	if err != nil {
		return nil, model.NewAppError("DoActionRequest", "api.post.do_action.action_integration.app_error", nil, "err="+err.Error(), http.StatusBadRequest)
	}
	r.Header.Set("Mattermost-User-ID", a.Session.UserId)
	r.Header.Set(model.HEADER_AUTH, "Bearer "+a.Session.Token)
	params := make(map[string]string)
	params["plugin_id"] = pluginId
	r = mux.SetURLVars(r, params)
	
	a.ServePluginRequest(w, r)

	resp := &http.Response{
		Status:           "",
		StatusCode:       w.status,
		Proto:            "HTTP/1.1",
		ProtoMajor:       1,
		ProtoMinor:       1,
		Header:           make(http.Header),
		Body:             ioutil.NopCloser(bytes.NewReader(w.data)),
		ContentLength:    0,
		TransferEncoding: nil,
		Close:            true,
		Uncompressed:     true,
		Trailer:          make(http.Header),
		Request:          nil,
		TLS:              nil,
	}

	return resp, nil
}
```

Reply:
It was a small bug that needed addition.  Fixed.